### PR TITLE
FINERACT-1971: Fix when waived N+1 installment's obligationsMet and obligationsMetOnDate is not or wrongly set

### PR DESCRIPTION
**Pull request by galovics**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java [hunk 1]
```diff
@@ -395,6 +397,16 @@ protected void reprocessInstallments(List<LoanRepaymentScheduleInstallment> inst
             if (loanRepaymentScheduleInstallment.getTotalOutstanding(currency).isGreaterThanZero()) {
                 loanRepaymentScheduleInstallment.updateObligationMet(false);
                 loanRepaymentScheduleInstallment.updateObligationMetOnDate(null);
+            } else {
+                Optional<LoanTransaction> optWaiverTx = transactions.stream().filter(lt -> {
+                    LocalDate fromDate = loanRepaymentScheduleInstallment.getFromDate();
+                    return lt.getSubmittedOnDate().isAfter(fromDate);
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java]: i think you need to use the transaction date here.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java [hunk 2]
```diff
@@ -395,6 +397,16 @@ protected void reprocessInstallments(List<LoanRepaymentScheduleInstallment> inst
             if (loanRepaymentScheduleInstallment.getTotalOutstanding(currency).isGreaterThanZero()) {
                 loanRepaymentScheduleInstallment.updateObligationMet(false);
                 loanRepaymentScheduleInstallment.updateObligationMetOnDate(null);
+            } else {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java]: please check whether the obligation met is `false` and there is no `obligationsMetDate` already set. We need to calculate only if thats the case

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java (hunk 1 above)
Comment (galovics) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java]: done.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java (hunk 2 above)
Comment (galovics) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/AbstractLoanRepaymentScheduleTransactionProcessor.java]: done.

**Review by adamsaghy — APPROVED** [review]
LGTM

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanWaiveChargeTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            } else {
+                product = createOnePeriod30DaysLongNoInterestPeriodicAccrualProduct();
+            }
+
+            PostLoanProductsResponse loanProductResponse = loanProductHelper.createLoanProduct(product);
+            Long loanProductId = loanProductResponse.getResourceId();
+
+            // Apply and Approve Loan
+
+            PostLoansRequest applicationRequest = applyLoanRequest(clientId, loanProductId, "01 January 2023", amount, 1);
```
Comment (reluxa) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanWaiveChargeTest.java:74]: https://github.com/apache/fineract/blob/9a5b8cb17b27866760d7d534e666f7e6278db182/integration-tests/src/test/java/org/apache/fineract/integrationtests/BaseLoanIntegrationTest.java#L371-L380

This overrides the payment allocation strategy created on the loan at the time of approval. So it'll always be: 
`DUE_PENALTY_INTEREST_PRINCIPAL_FEE_IN_ADVANCE_PENALTY_INTEREST_PRINCIPAL_FEE_STRATEGY`

Comment (galovics) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanWaiveChargeTest.java:74]: Nice Peter, nice.

Comment (galovics) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanWaiveChargeTest.java:74]: Fixed.

**Pull request merged**